activerecord/lib/active_record/encryption/encryptable_record.rb

```diff
@@ -125,7 +125,7 @@ def validate_column_size(attribute_name)
             if table_exists? && limit = columns_hash[attribute_name.to_s]&.limit
               validates_length_of attribute_name, maximum: limit
             end
-          rescue ActiveRecord::ConnectionNotEstablished => e
+          rescue ActiveRecord::ConnectionNotEstablished, ActiveRecord::DatabaseConnectionError => e
             Rails.logger.warn "Skipping adding length validation for #{self.name}\##{attribute_name}. Can't check column limit due to: #{e.inspect}"
           end
```

Comment:
Honestly I think this whole check should be reconsidered, having to access the DB at class definition time is kind of an anti-pattern. Can we check who introduced this and bring them in the discussion to see if this could either be eliminated or moved later in the boot process?

I'm thinking it could be part of `define_attribute_methods` as that's the time when we expect to have access to the database schema.

Reply:
Agreed... cc/ @jorgemanrubia would love to hear your thoughts?

Reply:
So the problem is because when you declare an attribute as encrypted with:

```
class Post < ActiveRecord::Base
    encrypts :title
end
```

Internally, it adds a validation that validates the length of the attribute based on the column size (see `validate_column_size` option in the [guide](https://edgeguides.rubyonrails.org/active_record_encryption.html)). To read the column size, it checks the database schema. All this happens at the "class declaration" level.

We can definitely run that logic ahead in the process, when a connection with the database happens. I'll investigate a good approach to do that, but if you have suggestions please shoot. Some kind of [callback](https://stackoverflow.com/questions/21446094/a-callback-for-activerecord-database-connections) would make sense to me.

Reply:
`load_schema!` could make sense: https://github.com/rails/rails/blob/c84dec3513581db444b2e055ccf79f54fdff4ecc/activerecord/lib/active_record/model_schema.rb#L563

Or as mentioned previously, `define_attribute_methods` as that's when Active Record generate all the methods based on the DB schema.

Reply:
Taking a stab here https://github.com/rails/rails/pull/43009. I think I prefer to go `load_schema!`, but maybe it's because I am not understanding how would you use `define_attribute_methods` here. We can move the discussion about the solution there. Thanks!